CMakeLists.txt

```diff
@@ -1,332 +1,335 @@
+option(NGEN_QUIET            "Silence output"                     OFF)
+
+# These options require dependency on some of the above options
+# Syntax: cmake_dependent_option(<option> "<help_text>" <value> <depends> <force>)
+cmake_dependent_option(NGEN_WITH_ROUTING "Build with t-route integration" ON "NGEN_WITH_PYTHON" OFF)
+cmake_dependent_option(NGEN_UPDATE_GIT_SUBMODULES "Update submodules on configure" ON "GIT_FOUND;NGEN_HAS_GIT_DIR" OFF)
+
+option(BMI_FORTRAN_ISO_C_LIB_DIR "Directory hint for middleware Fortran shared lib handling iso_c_binding" "${NGEN_ROOT_DIR}/extern/iso_c_fortran_bmi/cmake_build")
+option(BMI_FORTRAN_ISO_C_LIB_NAME "Name for middleware Fortran shared lib handling iso_c_binding" "iso_c_bmi")
+
+include(DeprecatedOptions)
+deprecated_option(OLD MPI_ACTIVE            NEW NGEN_WITH_MPI     INCLUDE_ENV)
+deprecated_option(OLD UDUNITS_ACTIVE        NEW NGEN_WITH_UDUNITS INCLUDE_ENV)
+deprecated_option(OLD NGEN_ACTIVATE_PYTHON  NEW NGEN_WITH_PYTHON  INCLUDE_ENV)
+deprecated_option(OLD NGEN_ACTIVATE_ROUTING NEW NGEN_WITH_ROUTING)
+deprecated_option(OLD NETCDF_ACTIVE         NEW NGEN_WITH_NETCDF)
+deprecated_option(OLD BMI_C_LIB_ACTIVE      NEW NGEN_WITH_BMI_C)
+deprecated_option(OLD BMI_FORTRAN_ACTIVE    NEW NGEN_WITH_BMI_FORTRAN)
+deprecated_option(OLD QUIET                 NEW NGEN_QUIET)
+deprecated_option(OLD PACKAGE_TESTS         NEW NGEN_WITH_TESTS)
+# Project =====================================================================
+
+# -----------------------------------------------------------------------------
+# Set global C++ options
+set(CMAKE_CXX_STANDARD 14)
+set(CMAKE_CXX_STANDARD_REQUIRED True)
+set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+
+# Define project version and use via generated config header
 project(ngen VERSION 0.1.0)
-configure_file(include/NGenConfig.h.in include/NGenConfig.h)
 
-if("${PROJECT_NAME}" STREQUAL "${CMAKE_PROJECT_NAME}")
-    option(PACKAGE_TESTS "Build automated tests" ON)
+configure_file("${NGEN_INC_DIR}/NGenConfig.h.in" "${NGEN_INC_DIR}/NGenConfig.h")
+
+add_executable(ngen "${NGEN_SRC_DIR}/NGen.cpp")
+
+# Dependencies ================================================================
+
+# -----------------------------------------------------------------------------
+# Check if NGen is the main project (e.g. development environment)
+if(PROJECT_NAME STREQUAL CMAKE_PROJECT_NAME)
+    set(NGEN_IS_MAIN_PROJECT ON)
 else()
-    option(PACKAGE_TESTS "Build automated tests")
+    set(NGEN_IS_MAIN_PROJECT OFF)
 endif()
 
+# -----------------------------------------------------------------------------
+# If MPI support is enabled, set the C++ compiler to "mpicxx"
+if(NGEN_WITH_MPI)
+    find_package(MPI REQUIRED)
+    add_compile_definitions(NGEN_MPI_ACTIVE)
+endif()
+
+# -----------------------------------------------------------------------------
+# FIXME: Is this needed for icpx? or even icpp? This only enables support for SYCL - Justin
```

Comment:
I added this when trying to get the build working on Hera... I don't remember exactly the symptoms, but I was having a hard time getting everything to use icpx, and adding this fixed it. This was a scenario where you had to load up the GCC9 module to get its stdlib, and then the OneAPI module to get icpx... but certain things (gfortran/mpif90?) kept going back to using GCC toolchain, and when I added this that problem stopped happening.... if I remember correctly (that was close to a year ago). It could be that something else is the "right" solution and this fixed it by accident, but I'm not sure what the right answer is then.

Reply:
Hmm maybe the IntelDPCPPConfig.cmake file is different on Hera -- the one I have doesn't modify anything besides SYCL variables. I also have to explicitly set `INTEL_DPCPP` for this branch to execute, even with `IntelLLVM` as my toolchain... so maybe it's an environment requirement. In any case, I don't think there's an issue with keeping this section of code in, so I'll modify the comment to provide context and leave it as is 😄

Reply:
I'll add (for posterity) that there is probably a smarter way of figuring out whether it ought to be turned on (i.e. turn on if CMAKE_CXX_COMPILER is icpx (which is non-trivial if it's `mpicxx`) or something) but I just kludged in a -D solution.

Reply:
"If CMAKE_CXX_COMPILER is icpx" would be a direct test of `if (CMAKE_CXX_COMPILER_ID STREQ IntelLLVM)`

This block is sketchy, but not really a big deal right now.

For context, the Intel compilers (and Nvidia, not that it matters for us) don't ship their own implementation of the standard library or the compiler-ABI-defining headers. Rather, they defer to a reference version of GCC either based on `PATH` or a command-line argument setting the reference explicitly to get their `libstdc++`. That's why they need the GCC9 module (for instance) loaded - it has to be a version that supports the language standard icpx is compiling for, but crucially, not one whose headers rely on language version support beyond what icpc implements. It's a really delicate dance, with poorly documented compatibility.